### PR TITLE
Config flow switches

### PR DESCRIPTION
**Pull request by andrewmeepos**

### CODE REVIEW

**Review by postlund — CHANGES_REQUESTED** [review]
I added a few comments and also pointed to where you can make some changes to adhere what I changed in my other PR. It will hopefully be a bit easier for me to read when you strip the debug printouts 😅

**Comment by andrewmeepos** [comment]
Thank! I'll be home in a couple hours and will incorporate these changes and remove the excessive debugging. 

This is my first experience working with Home Assistant (or, Python, for anything major), so the absurd amount of debug logs helped me follow exactly what was happening. 

-Andrew

File: custom_components/localtuya/config_flow.py [hunk 1]
```diff
@@ -175,11 +209,22 @@ def _convert_entity(conf):
         self._set_platform(user_input[CONF_PLATFORM])
 
         if len(user_input[CONF_SWITCHES]) > 0:
```
[the remark this answers is not in the document]
Comment (andrewmeepos) [reply custom_components/localtuya/config_flow.py:211]: What if someone has both a switch and a cover in their configuration.yaml? It won't import both in one call of `async def async_step_import(self, user_input)`? Would it do two calls of that function instead (one for the cover, one for the switch)?

I'll update it to one if elif else with len of the .get.

File: custom_components/localtuya/cover.py [hunk 2]
```diff
@@ -32,209 +71,422 @@
...
+    CONF_STOP_CMD,
+    CONF_GET_POSITION_KEY,
+    CONF_SET_POSITION_KEY,
+    DOMAIN,
+)
 
-DEFAULT_NAME = 'localtuyacover'
+_LOGGER = logging.getLogger(__name__)
 
 REQUIREMENTS = ['pytuya==7.0.9']
```
[the remark this answers is not in the document]
Comment (andrewmeepos) [reply custom_components/localtuya/cover.py:111]: I can't find the line `DEFAULT_NAME = 'localtuyacover` in the cover.py file of this commit...it's not in the file. i'm going to mark this as resolved.

File: custom_components/localtuya/config_flow.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (postlund) [reply custom_components/localtuya/config_flow.py:211]: Each platform is imported as it's own config entry, so that's not possible.

[the remark this answers is not in the document]
Comment (postlund) [reply custom_components/localtuya/config_flow.py:211]: Each platform is imported as it's own config entry, so that's not possible.

File: custom_components/localtuya/cover.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (postlund) [reply custom_components/localtuya/cover.py:111]: If it's not used you can remove it as well. But I was referring to `REQUIREMENTS` here. It's replaced by `manifest.json`.

File: custom_components/localtuya/config_flow.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewmeepos) [reply custom_components/localtuya/config_flow.py:211]: > Each platform is imported as it's own config entry, so that's not possible.

Ah, makes sense.

**Comment by andrewmeepos** [comment]
@postlund ok, I updated cover to address issues mentioned here, and prepped for the changes in the light_config_flow. I haven't had a chance to test this out yet. With the new schemas and helper functions, I don't think the necessary DPS keys are being properly passed to/from cover anymore (such as the key for set position and get position). I could fix this with some debugging, but i'm out of time for the night. Feel free to fix it yourself if you would like.

Also, in prepping for light_config_flow, I refactored cover.py to be more in line with pytuya __init__.py. For example, in cover.py, I now define:

```
    def open_cover(self):
        self._device.open_cover() 
```

And in pytuya, open_cover() is:
```
    def open_cover(self, switch=1):
        """Turn the device on"""
        self.set_status('on', switch)
```

My issue with this pytuya function, and some of the others, is that they don't account for the dps keys or other setup/config values. For example, on my cover device, my open cover command is "Open", not "on". "Open" is passed in through setup/config. 

Looking back, i'm not entirely sure why I refactored cover.py to be more in line with pytuya, other than I thought it would make for a cleaner codebase. However, i'm now second guessing what I did, as I think it's going to lead to more issues.

I think this leads us to a fork in the road as to how we want to proceed:
(1) Change some of cover.py's functions to use pytuya's more general set_value function (this is more or less what cover.py did before I refactored it)
(2) Update pytuya to allow for more extensibility/configurability in its defined commands (such as open_cover)

I think we probably want to go with (1).

Anyways, in summary, it's very possible that for every issue I solved with this pull request, I created a separate new issue. For that, I apologize. It's been a bit of a steep learning curve, but i'm starting to understand things better and better.

I'll have some time to work on things again tomorrow night. In the interim, feel free to fix as much of cover.py as you would like, or leave me with additional change requests and I can handle them tomorrow.

**Comment by postlund** [comment]
I made an attempt to implement cover support in:

https://github.com/rospogrigio/localtuya-homeassistant/pull/8

Didn't change any logic, just pure convert like I did with switch and light. Feel free to try it out, hopefully it works!

**Comment by andrewmeepos** [comment]
> I made an attempt to implement cover support in:
> 
> [rospogrigio#8](https://github.com/rospogrigio/localtuya-homeassistant/pull/8)
> 
> Didn't change any logic, just pure convert like I did with switch and light. Feel free to try it out, hopefully it works!

Thanks for this. It's close but needs a few tweaks. I've made most of the changes already. However, there is one last thing holding me up. Here's my updated PLATFORM_SCHEMA in cover.py:
```
PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(BASE_PLATFORM_SCHEMA).extend(
    {
        vol.Required(CONF_COVER_COMMAND_KEY, default=DEFAULT_COVER_COMMAND_KEY): cv.positive_int, # Need to modify to validate against vol.In(dps_strings), 
        vol.Required(CONF_SET_POSITION_KEY, default=DEFAULT_SET_POSITION_KEY): cv.positive_int, # Need to modify to validate against vol.In(dps_strings), 
        vol.Required(CONF_GET_POSITION_KEY, default=DEFAULT_GET_POSITION_KEY): cv.positive_int, # Need to modify to validate against vol.In(dps_strings),
        vol.Optional(CONF_OPEN_CMD, default=DEFAULT_OPEN_CMD): cv.string,
        vol.Optional(CONF_CLOSE_CMD, default=DEFAULT_CLOSE_CMD): cv.string,
        vol.Optional(CONF_STOP_CMD, default=DEFAULT_STOP_CMD): cv.string,
    }
)
```
I need to validate CONF_COVER_COMMAND_KEY, CONF_SET_POSITION_KEY and CONF_GET_POSITION_KEY against the dps data, just as is done for CONF_ID in config_flow.py:
```
            vol.Required(CONF_ID): **vol.In(dps_strings),**
```

However, I can't call `vol.In(dps_strings)` from cover.py,  as dps_keys/self.dps_keys is not defined in cover.py. I've tried doing various imports in cover.py from config_flow.py, but am having trouble getting it to work.

How can I import the dps strings into cover.py from config_flow.py so that I can validate against them for the PLATFORM_SCHEMA?

**Comment by postlund** [comment]
> > I made an attempt to implement cover support in:
> > [rospogrigio#8](https://github.com/rospogrigio/localtuya-homeassistant/pull/8)
> > Didn't change any logic, just pure convert like I did with switch and light. Feel free to try it out, hopefully it works!
> 
> Thanks for this. It's close but needs a few tweaks. I've made most of the changes already. However, there is one last thing holding me up. Here's my updated PLATFORM_SCHEMA in cover.py:
> 
> ```
> PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(BASE_PLATFORM_SCHEMA).extend(
>     {
>         vol.Required(CONF_COVER_COMMAND_KEY, default=DEFAULT_COVER_COMMAND_KEY): cv.positive_int, # Need to modify to validate against vol.In(dps_strings), 
>         vol.Required(CONF_SET_POSITION_KEY, default=DEFAULT_SET_POSITION_KEY): cv.positive_int, # Need to modify to validate against vol.In(dps_strings), 
>         vol.Required(CONF_GET_POSITION_KEY, default=DEFAULT_GET_POSITION_KEY): cv.positive_int, # Need to modify to validate against vol.In(dps_strings),
>         vol.Optional(CONF_OPEN_CMD, default=DEFAULT_OPEN_CMD): cv.string,
>         vol.Optional(CONF_CLOSE_CMD, default=DEFAULT_CLOSE_CMD): cv.string,
>         vol.Optional(CONF_STOP_CMD, default=DEFAULT_STOP_CMD): cv.string,
>     }
> )
> ```
> 
> I need to validate CONF_COVER_COMMAND_KEY, CONF_SET_POSITION_KEY and CONF_GET_POSITION_KEY against the dps data, just as is done for CONF_ID in config_flow.py:
> 
> ```
>             vol.Required(CONF_ID): vol.In(dps_strings),
> ```
> 
> However, I obviously can't call `vol.In(dps_strings)` from cover.py, as dps_strings/self.dps_strings is not defined in cover.py. I've tried doing various imports in cover.py from config_flow.py, but am having trouble getting it to work.
> 
> How can I import the dps strings into cover.py from config_flow.py so that I can validate against them for the PLATFORM_SCHEMA?

Simply put: you can't. The platform schema is static and everything about it must be known when defining it. We have to consider that we have two sides of the coin here. The config flow takes input from the user and creates a config entry. After that the config flow and everything about it is null and void. The code you are referring to here ("setup") takes a config entry and create entities from it. You can more or less compare a config entry with its corresponding YAML definition. In reality this means that we can do all sorts of validation before putting anything in the config entry. The platform setup code however has much less knowledge about dynamic parameters than config flows does. As mentioned, the schema is static and it's the same schema used for all entities that are setup. Since different devices can have different allowed DPS values, this would be a problem.

One thing to note here though. In the YAML config, if you accept a string the user can put anything there and have Home Assistant read that config. When it comes to config entries, Home Assistant and the config flow becomes the gate keeper of what goes into the config entry. So when setting up an entity from en config entry, you already know that the configuration is valid because the config flow would not allow anything else (assuming it does its job well). One of the benefits of supporting setup only via config flow.

**Comment by andrewmeepos** [comment]
@postlund all done! Works fantastic. Going to push any second.

**Comment by postlund** [comment]
Great! Now we just need to get some speed up in the main repo, so we can merge the PRs I have open. You can apply changes on top of that.

**Comment by postlund** [comment]
@postlund I wrote my two cents about it in my first PR, we'll see where we end up with that. Just keep track of the conversations. Nothing shall be merged to this repo though, it's only my local fork that I use to submit PRs.

1. It's probably fine, there are lot of improvements that can be made in all of the platforms. I have started small, but can make a lot of updates once my PRs are in. Don't want to mix to much updates in to the same PR, it makes it hard to follow and review. Hopefully we can fit everything together in the end.

2. Probably also fine as it aligns with the changes made to pytuya.

3. Looks good to me! 👍

4. You can validate against DPS values just like it's done with ID for instance. All the values are passed in to `flow_schema`, you can look at how it's done in `swich.py`. Super simple. You can, as previously mentioned, not have this check when using YAML though.

I say: let people enter what they want and if it doesn't work, it's up to them to figure it out. Don't try to add too much magic to free text fields.

Yeah, the way the protocol logic is implemented now is not very good. We will improve this over time, so no worries. Should be fixed at some point.

The entity itself has no idea that the methods alters the state of the entity, unless someone tells it about it or a scheduled update is made. So for that to propagate you would need to tell Home Assistant to force an early update of the entity and update state after stopping. You can do that with `async_update_ha_state`:

https://github.com/home-assistant/core/blob/062ac5f27d311c4935ef371ad31d07e6fc7ff249/homeassistant/helpers/entity.py#L260

**Comment by andrewmeepos** [comment]
> Nothing shall be merged to this repo though, it's only my local fork that I use to submit PRs.

Got it, i'll submit this as a pull request to rospogrigio and write that it depends on https://github.com/rospogrigio/localtuya-homeassistant/pull/6 and likely supersedes https://github.com/rospogrigio/localtuya-homeassistant/pull/8

**Comment by postlund** [comment]
I would turn it around a bit. Since your PR is essentially what's in rospogrigio#8 but with additional features, it should be enough to apply it after rospogrigio#8. This would yield a much smaller PR.

**Comment by andrewmeepos** [comment]
> I would turn it around a bit. Since your PR is essentially what's in [rospogrigio#8](https://github.com/rospogrigio/localtuya-homeassistant/pull/8) but with additional features, it should be enough to apply it after [rospogrigio#8](https://github.com/rospogrigio/localtuya-homeassistant/pull/8). This would yield a much smaller PR.

That makes sense, i'll clarify that in the PR
